{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddd2ecf8-514d-47a5-a547-77c875241754",
   "metadata": {},
   "outputs": [],
   "source": [
    "GOAL: go from handcrafted codon table to a final codon table\n",
    "\n",
    "# keys:\n",
    "- sele = selected missense variants\n",
    "- syn = includes synonymous codon variant\n",
    "- no_stop = removes missense stop codons\n",
    "- no_stop_syn = no stop codons AND includes synonymous codon\n",
    "\n",
    "# key columns:\n",
    "- codons list\n",
    "- aa string\n",
    "- iupac_codon (for final dictionary and variant codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a90ed58b-a68b-4906-81fb-e562de284ae7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add option for stop codon dictionary\n",
    "import pandas as pd\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7c17e0ea-0577-46f2-84ac-7b81e5b771e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "iupac_dict = {'A':'A','C':'C','G':'G','T':'T','AC':'M','AG':'R','AT':'W','CG':'S','CT':'Y','GT':'K','ACG':'V','ACT':'H','AGT':'D','CGT':'B','ACGT':'N'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "68d93cd0-bb28-4e7f-ad07-7636fe2093fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "rev_iupac_dict = {value:key for key,value in iupac_dict.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f1a6d9a5-01d3-4ee3-a659-72112b450848",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('dms_codon_table_v2.csv')\n",
    "df.fillna('', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6196dd69-e29c-425e-bb9f-741f67b0f749",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the sele_aa column is incorrect\n",
    "def sele_aa(row):\n",
    "    codons = row['sele_missense_codons'].split(' ')\n",
    "    aa_list = []\n",
    "    for codon in codons:\n",
    "        if codon == '':\n",
    "            return ''\n",
    "        else:\n",
    "            aa = str(Seq(codon).translate())\n",
    "            aa_list.append(aa)\n",
    "    return ''.join(aa_list)\n",
    "\n",
    "df['sele_aa'] = df.apply(sele_aa, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5c4ac2f8-d7e7-4273-8f02-221132c8e680",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_stop_codons(row):\n",
    "    codons = row['sele_missense_codons'].split(' ')\n",
    "    no_stop_codons = []\n",
    "    for codon in codons:\n",
    "        aa = str(Seq(codon).translate())\n",
    "        if aa != \"*\":\n",
    "            no_stop_codons.append(codon)\n",
    "    return ' '.join(no_stop_codons)\n",
    "\n",
    "df['no_stop_codons'] = df.apply(remove_stop_codons, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "eb62d010-6cb8-4af7-88b9-1ce4a866cedc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def no_stop_iupac(row):\n",
    "    codons = row['no_stop_codons'].split(' ')\n",
    "    #if not codons:\n",
    "        #return ''\n",
    "    pos = int(row['position'])\n",
    "    iupac = []\n",
    "    for codon in codons:\n",
    "        if codon == '':\n",
    "            return ''\n",
    "        else:\n",
    "            iupac.append(codon[pos])\n",
    "    iupac=set(iupac)\n",
    "    for key in iupac_dict.keys():\n",
    "        if set(list(key)) == iupac:\n",
    "            return iupac_dict[key]\n",
    "\n",
    "df['no_stop_iupac'] = df.apply(no_stop_iupac, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3f366b00-2b24-4cb5-9bcf-04333400681b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def no_stop_iupac_codon(row):\n",
    "    iupac = row['no_stop_iupac']\n",
    "    if iupac == '':\n",
    "        return ''\n",
    "    codon = row['codon']\n",
    "    pos = int(row['position'])\n",
    "    iupac_codon = codon[:pos] + iupac + codon[pos + 1:]\n",
    "    return iupac_codon\n",
    "\n",
    "df['no_stop_iupac_codon'] = df.apply(no_stop_iupac_codon, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3ddd656c-053c-429d-889f-edbff8ebe127",
   "metadata": {},
   "outputs": [],
   "source": [
    "def no_stop_nucleotides(row):\n",
    "    codons = row['no_stop_codons'].split(' ')\n",
    "    pos = int(row['position'])\n",
    "    nucs = []\n",
    "    for codon in codons:\n",
    "        if codon == '':\n",
    "            return ''\n",
    "        else:\n",
    "            nucs.append(codon[pos])\n",
    "    return ''.join(nucs)\n",
    "\n",
    "df['no_stop_nucleotides'] = df.apply(no_stop_nucleotides, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f87b8ac3-8914-4ec6-913b-1a1d57284a7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def no_stop_aa(row):\n",
    "    codons = row['no_stop_codons'].split(' ')\n",
    "    aa_list = []\n",
    "    for codon in codons:\n",
    "        if codon == '':\n",
    "            return ''\n",
    "        else:\n",
    "            aa = str(Seq(codon).translate())\n",
    "            aa_list.append(aa)\n",
    "    return ''.join(aa_list)\n",
    "\n",
    "df['no_stop_aa'] = df.apply(no_stop_aa, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9a839c5f-5b8b-4bcf-9c6c-fecee313a07e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     173\n",
       "*     19\n",
       "dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# return difference between stops and all aa\n",
    "def stop_aa_dif(row):\n",
    "    all_aa = set(list(row['sele_aa']))\n",
    "    stop_aa = set(list(row['no_stop_aa']))\n",
    "    diff = all_aa - stop_aa\n",
    "    return ''.join(diff)\n",
    "\n",
    "df.apply(stop_aa_dif, axis=1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ad8c4c00-b6cc-4858-99ff-fdd509116e5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# general functions\n",
    "def get_iupac_symbol(nuc_set):\n",
    "    \"\"\"Return IUPAC symbol for a set of nucleotides\"\"\"\n",
    "    iupac_dict = {'A':'A','C':'C','G':'G','T':'T','AC':'M','AG':'R','AT':'W','CG':'S','CT':'Y','GT':'K','ACG':'V','ACT':'H','AGT':'D','CGT':'B','ACGT':'N'}\n",
    "    for key in iupac_dict.keys():\n",
    "        if set(key) == nuc_set:\n",
    "            return iupac_dict[key]\n",
    "\n",
    "# make a new aa column from codons list\n",
    "def make_aa_col(missense_codons_col):\n",
    "    codons = missense_codons_col.split(' ')\n",
    "    aa_list = [str(Seq(codon).translate()) for codon in codons]\n",
    "    return ''.join(aa_list)\n",
    "\n",
    "# make a new iupac_codon column from codons list\n",
    "def make_iupac_codon_col(wt_col, pos_col, missense_codons_col):\n",
    "    codons = missense_codons_col.split(' ')\n",
    "    if '' in codons:\n",
    "        return ''\n",
    "    wt_codon = str(wt_col)\n",
    "    pos = int(pos_col)\n",
    "    nucs = {codon[pos] for codon in codons}\n",
    "    iupac = get_iupac_symbol(nucs)\n",
    "    return wt_codon[:pos] + iupac + wt_codon[pos + 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6424b713-2d2d-40d6-acb0-e35d49a86412",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>codon</th>\n",
       "      <th>aa</th>\n",
       "      <th>position</th>\n",
       "      <th>all_missense_aa</th>\n",
       "      <th>all_missense_nucleotides</th>\n",
       "      <th>all_missense_codons</th>\n",
       "      <th>all_iupac</th>\n",
       "      <th>all_iupac_codon</th>\n",
       "      <th>sele_missense_codons</th>\n",
       "      <th>syn_iupac_codon</th>\n",
       "      <th>...</th>\n",
       "      <th>sele_aa</th>\n",
       "      <th>syn_aa</th>\n",
       "      <th>syn_missense_codons</th>\n",
       "      <th>no_stop_codons</th>\n",
       "      <th>no_stop_iupac</th>\n",
       "      <th>no_stop_iupac_codon</th>\n",
       "      <th>no_stop_nucleotides</th>\n",
       "      <th>no_stop_aa</th>\n",
       "      <th>stop_syn_codons</th>\n",
       "      <th>stop_syn_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>0</td>\n",
       "      <td>*QE</td>\n",
       "      <td>CGT</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>B</td>\n",
       "      <td>BAA</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>BAA</td>\n",
       "      <td>...</td>\n",
       "      <td>QE*</td>\n",
       "      <td>*QE</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>CAA GAA</td>\n",
       "      <td>S</td>\n",
       "      <td>SAA</td>\n",
       "      <td>CG</td>\n",
       "      <td>QE</td>\n",
       "      <td>CAA GAA</td>\n",
       "      <td>QE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>1</td>\n",
       "      <td>TRI</td>\n",
       "      <td>CGT</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>B</td>\n",
       "      <td>ABA</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>ABA</td>\n",
       "      <td>...</td>\n",
       "      <td>TRI</td>\n",
       "      <td>RTI</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>B</td>\n",
       "      <td>ABA</td>\n",
       "      <td>CGT</td>\n",
       "      <td>TRI</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>TRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>CT</td>\n",
       "      <td>AAC AAT</td>\n",
       "      <td>Y</td>\n",
       "      <td>AAY</td>\n",
       "      <td>AAT</td>\n",
       "      <td>AAK</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>KN</td>\n",
       "      <td>AAG AAT</td>\n",
       "      <td>AAT</td>\n",
       "      <td>T</td>\n",
       "      <td>AAT</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAC</td>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "      <td>HDY</td>\n",
       "      <td>CGT</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>B</td>\n",
       "      <td>BAC</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>BAC</td>\n",
       "      <td>...</td>\n",
       "      <td>HDY</td>\n",
       "      <td>DYH</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>B</td>\n",
       "      <td>BAC</td>\n",
       "      <td>CGT</td>\n",
       "      <td>HDY</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>HDY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAC</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>TSI</td>\n",
       "      <td>CGT</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>B</td>\n",
       "      <td>ABC</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>ABC</td>\n",
       "      <td>...</td>\n",
       "      <td>TSI</td>\n",
       "      <td>STI</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>B</td>\n",
       "      <td>ABC</td>\n",
       "      <td>CGT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>TSI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  codon aa  position all_missense_aa all_missense_nucleotides  \\\n",
       "0   AAA  K         0             *QE                      CGT   \n",
       "1   AAA  K         1             TRI                      CGT   \n",
       "2   AAA  K         2               N                       CT   \n",
       "3   AAC  N         0             HDY                      CGT   \n",
       "4   AAC  N         1             TSI                      CGT   \n",
       "\n",
       "  all_missense_codons all_iupac all_iupac_codon sele_missense_codons  \\\n",
       "0         CAA GAA TAA         B             BAA          CAA GAA TAA   \n",
       "1         ACA AGA ATA         B             ABA          ACA AGA ATA   \n",
       "2             AAC AAT         Y             AAY                  AAT   \n",
       "3         CAC GAC TAC         B             BAC          CAC GAC TAC   \n",
       "4         ACC AGC ATC         B             ABC          ACC AGC ATC   \n",
       "\n",
       "  syn_iupac_codon  ... sele_aa syn_aa syn_missense_codons no_stop_codons  \\\n",
       "0             BAA  ...     QE*    *QE         CAA GAA TAA        CAA GAA   \n",
       "1             ABA  ...     TRI    RTI         ACA AGA ATA    ACA AGA ATA   \n",
       "2             AAK  ...       N     KN             AAG AAT            AAT   \n",
       "3             BAC  ...     HDY    DYH         CAC GAC TAC    CAC GAC TAC   \n",
       "4             ABC  ...     TSI    STI         ACC AGC ATC    ACC AGC ATC   \n",
       "\n",
       "  no_stop_iupac no_stop_iupac_codon no_stop_nucleotides no_stop_aa  \\\n",
       "0             S                 SAA                  CG         QE   \n",
       "1             B                 ABA                 CGT        TRI   \n",
       "2             T                 AAT                   T          N   \n",
       "3             B                 BAC                 CGT        HDY   \n",
       "4             B                 ABC                 CGT        TSI   \n",
       "\n",
       "  stop_syn_codons stop_syn_aa  \n",
       "0         CAA GAA          QE  \n",
       "1     ACA AGA ATA         TRI  \n",
       "2             AAT           N  \n",
       "3     CAC GAC TAC         HDY  \n",
       "4     ACC AGC ATC         TSI  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3ee1409a-f1db-4e35-81cc-e64e3f7cd64d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['codon',\n",
       " 'aa',\n",
       " 'position',\n",
       " 'all_missense_aa',\n",
       " 'all_missense_nucleotides',\n",
       " 'all_missense_codons',\n",
       " 'all_iupac',\n",
       " 'all_iupac_codon',\n",
       " 'sele_missense_codons',\n",
       " 'syn_iupac_codon',\n",
       " 'syn_codon',\n",
       " 'sele_missense_nucleotides',\n",
       " 'sele_iupac',\n",
       " 'sele_iupac_codon',\n",
       " 'sele_aa',\n",
       " 'syn_aa',\n",
       " 'syn_missense_codons',\n",
       " 'no_stop_codons',\n",
       " 'no_stop_iupac',\n",
       " 'no_stop_iupac_codon',\n",
       " 'no_stop_nucleotides',\n",
       " 'no_stop_aa',\n",
       " 'stop_syn_codons',\n",
       " 'stop_syn_aa']"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "7a5d4924-2006-4c0b-9907-b9306d08060e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      SAA\n",
       "1      ABA\n",
       "2      AAT\n",
       "3      BAC\n",
       "4      ABC\n",
       "      ... \n",
       "187    TSG\n",
       "188    TTT\n",
       "189    RTT\n",
       "190    TVT\n",
       "191    TTG\n",
       "Length: 192, dtype: object"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make new stop_syn_codons column\n",
    "def stop_syn_codons(sele_codons_column, wt_aa_column):\n",
    "    codons = sele_codons_column.split(' ')\n",
    "    wt_aa = str(wt_aa_column)\n",
    "    for codon in codons:\n",
    "        # check if synonymous or stop codon\n",
    "        if str(Seq(codon).translate()) in [wt_aa,'*']:\n",
    "            codons.remove(codon)\n",
    "    return ' '.join(codons)\n",
    "\n",
    "df['no_stop_syn_codons'] = df.apply(lambda x: stop_syn_codons(x['sele_missense_codons'], x['aa']), axis=1)\n",
    "df['no_stop_syn_aa'] = df.apply(lambda x: make_aa_col(x['stop_syn_codons']), axis=1)\n",
    "df.apply(lambda x: make_iupac_codon_col(x['codon'], x['position'], x['stop_syn_codons']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f8698672-e1b4-4a94-a240-6208617b2681",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>codon</th>\n",
       "      <th>aa</th>\n",
       "      <th>position</th>\n",
       "      <th>all_missense_aa</th>\n",
       "      <th>all_missense_nucleotides</th>\n",
       "      <th>all_missense_codons</th>\n",
       "      <th>all_iupac</th>\n",
       "      <th>all_iupac_codon</th>\n",
       "      <th>sele_missense_codons</th>\n",
       "      <th>syn_iupac_codon</th>\n",
       "      <th>...</th>\n",
       "      <th>sele_aa</th>\n",
       "      <th>syn_aa</th>\n",
       "      <th>syn_missense_codons</th>\n",
       "      <th>no_stop_codons</th>\n",
       "      <th>no_stop_iupac</th>\n",
       "      <th>no_stop_iupac_codon</th>\n",
       "      <th>no_stop_nucleotides</th>\n",
       "      <th>no_stop_aa</th>\n",
       "      <th>stop_syn_codons</th>\n",
       "      <th>stop_syn_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>0</td>\n",
       "      <td>*QE</td>\n",
       "      <td>CGT</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>B</td>\n",
       "      <td>BAA</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>BAA</td>\n",
       "      <td>...</td>\n",
       "      <td>QE*</td>\n",
       "      <td>*QE</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>CAA GAA</td>\n",
       "      <td>S</td>\n",
       "      <td>SAA</td>\n",
       "      <td>CG</td>\n",
       "      <td>QE</td>\n",
       "      <td>CAA GAA</td>\n",
       "      <td>QE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>1</td>\n",
       "      <td>TRI</td>\n",
       "      <td>CGT</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>B</td>\n",
       "      <td>ABA</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>ABA</td>\n",
       "      <td>...</td>\n",
       "      <td>TRI</td>\n",
       "      <td>RTI</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>B</td>\n",
       "      <td>ABA</td>\n",
       "      <td>CGT</td>\n",
       "      <td>TRI</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>TRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>CT</td>\n",
       "      <td>AAC AAT</td>\n",
       "      <td>Y</td>\n",
       "      <td>AAY</td>\n",
       "      <td>AAT</td>\n",
       "      <td>AAK</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>KN</td>\n",
       "      <td>AAG AAT</td>\n",
       "      <td>AAT</td>\n",
       "      <td>T</td>\n",
       "      <td>AAT</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAC</td>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "      <td>HDY</td>\n",
       "      <td>CGT</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>B</td>\n",
       "      <td>BAC</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>BAC</td>\n",
       "      <td>...</td>\n",
       "      <td>HDY</td>\n",
       "      <td>DYH</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>B</td>\n",
       "      <td>BAC</td>\n",
       "      <td>CGT</td>\n",
       "      <td>HDY</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>HDY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAC</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>TSI</td>\n",
       "      <td>CGT</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>B</td>\n",
       "      <td>ABC</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>ABC</td>\n",
       "      <td>...</td>\n",
       "      <td>TSI</td>\n",
       "      <td>STI</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>B</td>\n",
       "      <td>ABC</td>\n",
       "      <td>CGT</td>\n",
       "      <td>TSI</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>TSI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  codon aa  position all_missense_aa all_missense_nucleotides  \\\n",
       "0   AAA  K         0             *QE                      CGT   \n",
       "1   AAA  K         1             TRI                      CGT   \n",
       "2   AAA  K         2               N                       CT   \n",
       "3   AAC  N         0             HDY                      CGT   \n",
       "4   AAC  N         1             TSI                      CGT   \n",
       "\n",
       "  all_missense_codons all_iupac all_iupac_codon sele_missense_codons  \\\n",
       "0         CAA GAA TAA         B             BAA          CAA GAA TAA   \n",
       "1         ACA AGA ATA         B             ABA          ACA AGA ATA   \n",
       "2             AAC AAT         Y             AAY                  AAT   \n",
       "3         CAC GAC TAC         B             BAC          CAC GAC TAC   \n",
       "4         ACC AGC ATC         B             ABC          ACC AGC ATC   \n",
       "\n",
       "  syn_iupac_codon  ... sele_aa syn_aa syn_missense_codons no_stop_codons  \\\n",
       "0             BAA  ...     QE*    *QE         CAA GAA TAA        CAA GAA   \n",
       "1             ABA  ...     TRI    RTI         ACA AGA ATA    ACA AGA ATA   \n",
       "2             AAK  ...       N     KN             AAG AAT            AAT   \n",
       "3             BAC  ...     HDY    DYH         CAC GAC TAC    CAC GAC TAC   \n",
       "4             ABC  ...     TSI    STI         ACC AGC ATC    ACC AGC ATC   \n",
       "\n",
       "  no_stop_iupac no_stop_iupac_codon no_stop_nucleotides no_stop_aa  \\\n",
       "0             S                 SAA                  CG         QE   \n",
       "1             B                 ABA                 CGT        TRI   \n",
       "2             T                 AAT                   T          N   \n",
       "3             B                 BAC                 CGT        HDY   \n",
       "4             B                 ABC                 CGT        TSI   \n",
       "\n",
       "  stop_syn_codons stop_syn_aa  \n",
       "0         CAA GAA          QE  \n",
       "1     ACA AGA ATA         TRI  \n",
       "2             AAT           N  \n",
       "3     CAC GAC TAC         HDY  \n",
       "4     ACC AGC ATC         TSI  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "135163fb-7db2-49a7-a3c3-f12aea1e752f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       QE\n",
       "1      TRI\n",
       "2        N\n",
       "3      HDY\n",
       "4      TSI\n",
       "      ... \n",
       "187     SW\n",
       "188      F\n",
       "189     IV\n",
       "190    YSC\n",
       "191      L\n",
       "Length: 192, dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "df['stop_syn_aa'] = df.apply(lambda x: make_aa_col(x['stop_syn_codons']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8a848090-32bb-4387-a170-f70327c14abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a new aa column from codons list\n",
    "def make_iupac_codon_col(missense_codons_col, pos_col):\n",
    "    codons = missense_codons_col.split(' ')\n",
    "    pos = int(pos_col)\n",
    "    nucs = {codon[pos] for codon in codons}\n",
    "    iupac = get_iupac_symbol(nucs)\n",
    "    return codon[:pos] + iupac + codon[pos + 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17305200-126e-45d4-bf1f-d154ad0ed6c8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "99b874f2-b166-465b-adca-e06ea642e573",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check number of missense stops\n",
    "df['sele_aa'].str.contains('\\*').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87ad8273-5fef-47db-a515-b08cd50e1677",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>codon</th>\n",
       "      <th>aa</th>\n",
       "      <th>position</th>\n",
       "      <th>all_missense_aa</th>\n",
       "      <th>all_missense_nucleotides</th>\n",
       "      <th>all_missense_codons</th>\n",
       "      <th>all_iupac</th>\n",
       "      <th>all_iupac_codon</th>\n",
       "      <th>sele_missense_codons</th>\n",
       "      <th>syn_iupac_codon</th>\n",
       "      <th>...</th>\n",
       "      <th>sele_iupac</th>\n",
       "      <th>sele_iupac_codon</th>\n",
       "      <th>sele_aa</th>\n",
       "      <th>syn_aa</th>\n",
       "      <th>syn_missense_codons</th>\n",
       "      <th>no_stop_codons</th>\n",
       "      <th>no_stop_iupac</th>\n",
       "      <th>no_stop_iupac_codon</th>\n",
       "      <th>no_stop_nucleotides</th>\n",
       "      <th>no_stop_aa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>0</td>\n",
       "      <td>*QE</td>\n",
       "      <td>CGT</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>B</td>\n",
       "      <td>BAA</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>BAA</td>\n",
       "      <td>...</td>\n",
       "      <td>B</td>\n",
       "      <td>BAA</td>\n",
       "      <td>QE*</td>\n",
       "      <td>*QE</td>\n",
       "      <td>CAA GAA TAA</td>\n",
       "      <td>CAA GAA</td>\n",
       "      <td>S</td>\n",
       "      <td>SAA</td>\n",
       "      <td>CG</td>\n",
       "      <td>QE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>1</td>\n",
       "      <td>TRI</td>\n",
       "      <td>CGT</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>B</td>\n",
       "      <td>ABA</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>ABA</td>\n",
       "      <td>...</td>\n",
       "      <td>B</td>\n",
       "      <td>ABA</td>\n",
       "      <td>TRI</td>\n",
       "      <td>RTI</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>ACA AGA ATA</td>\n",
       "      <td>B</td>\n",
       "      <td>ABA</td>\n",
       "      <td>CGT</td>\n",
       "      <td>TRI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAA</td>\n",
       "      <td>K</td>\n",
       "      <td>2</td>\n",
       "      <td>N</td>\n",
       "      <td>CT</td>\n",
       "      <td>AAC AAT</td>\n",
       "      <td>Y</td>\n",
       "      <td>AAY</td>\n",
       "      <td>AAT</td>\n",
       "      <td>AAK</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>AAT</td>\n",
       "      <td>N</td>\n",
       "      <td>KN</td>\n",
       "      <td>AAG AAT</td>\n",
       "      <td>AAT</td>\n",
       "      <td>T</td>\n",
       "      <td>AAT</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAC</td>\n",
       "      <td>N</td>\n",
       "      <td>0</td>\n",
       "      <td>HDY</td>\n",
       "      <td>CGT</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>B</td>\n",
       "      <td>BAC</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>BAC</td>\n",
       "      <td>...</td>\n",
       "      <td>B</td>\n",
       "      <td>BAC</td>\n",
       "      <td>HDY</td>\n",
       "      <td>DYH</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>CAC GAC TAC</td>\n",
       "      <td>B</td>\n",
       "      <td>BAC</td>\n",
       "      <td>CGT</td>\n",
       "      <td>HDY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAC</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>TSI</td>\n",
       "      <td>CGT</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>B</td>\n",
       "      <td>ABC</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>ABC</td>\n",
       "      <td>...</td>\n",
       "      <td>B</td>\n",
       "      <td>ABC</td>\n",
       "      <td>TSI</td>\n",
       "      <td>STI</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>ACC AGC ATC</td>\n",
       "      <td>B</td>\n",
       "      <td>ABC</td>\n",
       "      <td>CGT</td>\n",
       "      <td>TSI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  codon aa  position all_missense_aa all_missense_nucleotides  \\\n",
       "0   AAA  K         0             *QE                      CGT   \n",
       "1   AAA  K         1             TRI                      CGT   \n",
       "2   AAA  K         2               N                       CT   \n",
       "3   AAC  N         0             HDY                      CGT   \n",
       "4   AAC  N         1             TSI                      CGT   \n",
       "\n",
       "  all_missense_codons all_iupac all_iupac_codon sele_missense_codons  \\\n",
       "0         CAA GAA TAA         B             BAA          CAA GAA TAA   \n",
       "1         ACA AGA ATA         B             ABA          ACA AGA ATA   \n",
       "2             AAC AAT         Y             AAY                  AAT   \n",
       "3         CAC GAC TAC         B             BAC          CAC GAC TAC   \n",
       "4         ACC AGC ATC         B             ABC          ACC AGC ATC   \n",
       "\n",
       "  syn_iupac_codon  ... sele_iupac sele_iupac_codon sele_aa syn_aa  \\\n",
       "0             BAA  ...          B              BAA     QE*    *QE   \n",
       "1             ABA  ...          B              ABA     TRI    RTI   \n",
       "2             AAK  ...          T              AAT       N     KN   \n",
       "3             BAC  ...          B              BAC     HDY    DYH   \n",
       "4             ABC  ...          B              ABC     TSI    STI   \n",
       "\n",
       "  syn_missense_codons no_stop_codons no_stop_iupac no_stop_iupac_codon  \\\n",
       "0         CAA GAA TAA        CAA GAA             S                 SAA   \n",
       "1         ACA AGA ATA    ACA AGA ATA             B                 ABA   \n",
       "2             AAG AAT            AAT             T                 AAT   \n",
       "3         CAC GAC TAC    CAC GAC TAC             B                 BAC   \n",
       "4         ACC AGC ATC    ACC AGC ATC             B                 ABC   \n",
       "\n",
       "  no_stop_nucleotides no_stop_aa  \n",
       "0                  CG         QE  \n",
       "1                 CGT        TRI  \n",
       "2                   T          N  \n",
       "3                 CGT        HDY  \n",
       "4                 CGT        TSI  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('dms_no_stops.csv')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d7eb0a2-08df-4e05-8eae-eeced158ae2c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
