{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83691daa-92cf-4438-ba48-d4eeb44cc597",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make missense codon tables\n",
    "\n",
    "# this uses a corrected function to make missense variants\n",
    "\n",
    "from itertools import product\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "\n",
    "iupac_dict = {'A':'A','C':'C','G':'G','T':'T','AC':'M','AG':'R','AT':'W','CG':'S','CT':'Y','GT':'K','ACG':'V','ACT':'H','AGT':'D','CGT':'B','ACGT':'N'}\n",
    "\n",
    "\n",
    "# AA LEVEL (64 entries)\n",
    "nucleotides = 'ACGT'\n",
    "\n",
    "codon_list = []\n",
    "aa_list = []\n",
    "iupac_codon_list = []\n",
    "missense_aa_list = []\n",
    "\n",
    "\n",
    "for codon in list(''.join(w) for w in product(nucleotides, repeat=3)):\n",
    "    aa = Seq(codon).translate(table='Standard')[0]\n",
    "    new_aas = []\n",
    "    iupac_codon = []\n",
    "    \n",
    "    # loop through each position in codon\n",
    "    for position in range(3):\n",
    "        iupac_n = []\n",
    "        \n",
    "        # looping through each nucleotide\n",
    "        for n in nucleotides:\n",
    "            new_codon = codon[:position] + n + codon[position + 1:]\n",
    "            new_aa = Seq(new_codon).translate(table='Standard')[0]\n",
    "            if new_aa != aa:\n",
    "                new_aas.append(new_aa)\n",
    "                iupac_n.append(n)\n",
    "            else:\n",
    "                continue\n",
    "        \n",
    "        # iupac symbols\n",
    "        if not iupac_n: # check if the list is empty\n",
    "            iupac_codon.append(codon[position])\n",
    "        else:\n",
    "            for i in iupac_dict.keys():\n",
    "                if set(i) == set(iupac_n):\n",
    "                    iupac_codon.append(iupac_dict[i])\n",
    "    \n",
    "    # make assignments here\n",
    "    codon_list.append(codon)\n",
    "    aa_list.append(aa)\n",
    "    iupac_codon_list.append(''.join(iupac_codon))\n",
    "    missense_aa_list.append(''.join(set(new_aas)))\n",
    "\n",
    "# create AA Level dictionary\n",
    "aa_level_dict = {\n",
    "    'codon':codon_list, \n",
    "    'aa':aa_list, \n",
    "    'iupac_codon':iupac_codon_list,\n",
    "    'missense_aa':missense_aa_list   \n",
    "}\n",
    "\n",
    "df = pd.DataFrame.from_dict(aa_level_dict)\n",
    "df.to_csv('data/aa_missense_table.csv', index=False)\n",
    "\n",
    "# CODON LEVEL\n",
    "# df: codon, aa, position, nucleotides, missense_codons, missense_aa, iupac, iupac_codon\n",
    "nucleotides = 'ACGT'\n",
    "\n",
    "codon_list = []\n",
    "aa_list = []\n",
    "position_list = []\n",
    "nucleotides_list = []\n",
    "missense_codons_list = []\n",
    "missense_aa_list = []\n",
    "iupac_list = []\n",
    "iupac_codon_list = []\n",
    "\n",
    "for codon in list(''.join(w) for w in product(nucleotides, repeat=3)):\n",
    "    aa = Seq(codon).translate(table='Standard')[0]\n",
    "    \n",
    "    # loop through each position in codon\n",
    "    for position in range(3):      \n",
    "        new_aas = []\n",
    "        iupac_n = []\n",
    "        new_codons = []\n",
    "        # looping through each nucleotide\n",
    "        for i in nucleotides:\n",
    "            new_codon = codon[:position] + i + codon[position + 1:]\n",
    "            new_aa = Seq(new_codon).translate(table='Standard')[0]\n",
    "            if not new_aa == aa:\n",
    "                new_aas.append(new_aa)\n",
    "                iupac_n.append(i)\n",
    "                new_codons.append(new_codon)\n",
    "            else:\n",
    "                continue\n",
    "        \n",
    "        #iupac\n",
    "        if not iupac_n: # check if iupac_n is empty\n",
    "            iupac = codon[position]\n",
    "        else:\n",
    "            for i in iupac_dict.keys():\n",
    "                if set(i) == set(iupac_n):\n",
    "                    iupac = iupac_dict[i]\n",
    "      \n",
    "        # make assignments here\n",
    "        codon_list.append(codon)\n",
    "        aa_list.append(aa)\n",
    "        position_list.append(position)\n",
    "        nucleotides_list.append(''.join(iupac_n))\n",
    "        missense_codons_list.append(' '.join(new_codons))\n",
    "        missense_aa_list.append(''.join(set(new_aas)))\n",
    "        \n",
    "        \n",
    "        for i in iupac_dict.keys():\n",
    "            if set(i) == set(iupac_n):\n",
    "                iupac = iupac_dict[i]\n",
    "        iupac_list.append(iupac)\n",
    "        iupac_codon_list.append(codon[:position] + iupac + codon[position + 1:])\n",
    "\n",
    "# create dictionary\n",
    "codon_level_dict = {\n",
    "    'codon':codon_list, \n",
    "    'aa':aa_list, \n",
    "    'position':position_list,\n",
    "    'missense_nucleotides':nucleotides_list,\n",
    "    'missense_codons':missense_codons_list,\n",
    "    'missense_aa':missense_aa_list,\n",
    "    'iupac':iupac_list,\n",
    "    'iupac_codon':iupac_codon_list\n",
    "}\n",
    "\n",
    "df = pd.DataFrame.from_dict(codon_level_dict)\n",
    "df.to_csv('data/codon_missense_table.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
